### ISSUE TITLE
Kata not handling sizeLimit for memory backed ephemeral volumes 

### ISSUE THREAD

**Issue by bpradipt** [kata-containers/kata-containers#3765]
# Description of problem

Creating the following POD results in `/var/lib/containers`  to have a size which is 50% of VM memory. This is default behaviour when using tmpfs on Linux. However k8s 1.21+ introduced the `SizeMemoryBackedVolumes` feature gate which allows to set the size of memory backed ephemeral volume. The following K8s KEP describes the feature -https://github.com/kubernetes/enhancements/pull/1968

```
apiVersion: v1
kind: Pod
metadata:
  name: buildah-emptydir
spec:
  containers:
    - name: buildah
      image: quay.io/buildah/stable:v1.23.1
      command: ["sleep", "infinity"]      
      volumeMounts:
      - mountPath: /var/lib/containers
        name: container-storage
  runtimeClassName: kata
  volumes:
  - name: container-storage
    emptyDir:
      medium: Memory
      sizeLimit: 500M
```

# Expected result

```
[root@buildah-emptydir /]# df -h
Filesystem      Size  Used Avail Use% Mounted on
tmpfs           500M     0  500M   0% /var/lib/containers
```

# Actual result
```
[root@buildah-emptydir /]# df -h
Filesystem      Size  Used Avail Use% Mounted on
tmpfs           998M     0  998M   0% /var/lib/containers
```
This is approx 50% of the total VM memory (2G)

# Further information

**Comment by liubin** [comment]
Hi @bpradipt, FYI:

I had worked on a PoC version https://github.com/kata-containers/kata-containers/pull/2442, it works with basic usage.

In this PoC I hadn't found a way to get the size other than from mount info.

**Comment by boombatower** [comment]
This seems like the best/only practical workaround for the container build issues and generally useful on it's own.

Is this something that has a path forward or pending a decision on direction?

> As comments from others, there is still not an ideal solution now.

Otherwise, it doesn't seem clear what the underlying cause of the incompatibility is and if it can be resolved. From testing, it seems fickle in that buildx builds will work in some cases, and fail consistently in others. Even two steps doing effectively the same thing will consistently fail on only 1 of them.

Always works on tmpfs, assuming you have memory overhead. 1G is a little tight for many applications.

Short of that, I'm assuming this would also workaround with direct block devices? At which point, practically, need a way to allocate them dynamically or potentially from a pool of devices.

**Comment by bpradipt** [comment]
> This is something that I've hit in the context of:
> 
> * [Running the Engine in kata-containers fails with `failed to create hash for : operation not supported` dagger/dagger#9319](https://github.com/dagger/dagger/issues/9319)
> 
> I am currently looking at `csi-kata-directvolume` as a workaround. cc [@Apokleos](https://github.com/Apokleos)
> 
> What are your current thoughts on `sizeLimit` support in the context of kata-cotnainers [@liubin](https://github.com/liubin) [@fidencio](https://github.com/fidencio) ?
> 
> Have you made any progress / worked around this [@bpradipt](https://github.com/bpradipt) ?

@gerhard no. I didn't get time to work on this. For now I remount using a different size (eg. mount -o remount,size=3G /var/lib/containers)

**Comment by gmintoco** [comment]
I recently ran up against this, not sure if it is expected or not. But in my case, the emptyDir memory backed volume is always 1GB regardless of the memory limits/requests (and of course doesn't honor the sizeLimit parameter).

I am using the guidance from running docker in kata containers about mounting a emptydir volume for docker contents, of course 1GB is pretty minimal for that use case. Currently I am doing something similar to above where I remount the volume before starting dockerd.

**Comment by liubocflt** [comment]
Also run into this, the sizeLimit is not honored in kata container.

**Comment by egernst** [comment]
Duplicate of this one: https://github.com/kata-containers/kata-containers/issues/3765

**Comment by zachspar** [comment]
@egernst I am still seeing this problem. I'm running kata 3.12 using kata-qemu runtime class. And my tmpfs emptydir isn't able to go beyond 1GB in size, even though I'm setting size limit to 15GB. I am setting medium=memory, however this is not honored.

**Comment by c3d** [comment]
Hi @egernst I'm curious why this was closed. I don't see any reference to a commit or PR fixing this. If you don't mind, I will reopen for now.

**Comment by c3d** [comment]
I think that the problem was just the issue number in your comment, you probably meant #2438 and not #3765 (which is this one) 😉 

OK, closing again as duplicate of #2438
